addons/hr_holidays/models/hr_leave_allocation.py

```diff
@@ -674,7 +674,23 @@ def write(self, values):
         employee_id = values.get('employee_id', False)
         if values.get('state'):
             self._check_approval_update(values['state'])
+
+        previous_consumed_leaves = self.employee_id._get_consumed_leaves(leave_types=self.holiday_status_id)
         result = super(HolidaysAllocation, self).write(values)
+
+        # Check constraints after write to consider updated values.
+        if 'number_of_days_display' in values or 'number_of_hours_display' in values:
+            consumed_leaves = self.employee_id._get_consumed_leaves(leave_types=self.holiday_status_id)
+            for allocation in self:
+                current_excess_days = dict(consumed_leaves[1]).get(allocation.employee_id, {}).get(allocation.holiday_status_id, {}).get('excess_days', {})
+                previous_excess_days = dict(previous_consumed_leaves[1]).get(allocation.employee_id, {}).get(allocation.holiday_status_id, {}).get('excess_days', {})
+
+                total_current_excess_days = sum(map(lambda leave_date: leave_date['amount'], current_excess_days.values()))
+                total_previous_excess_days = sum(map(lambda leave_date: leave_date['amount'], previous_excess_days.values()))
+
+                if total_current_excess_days > total_previous_excess_days:
+                    raise ValidationError(_('You cannot reduce the duration below the duration of leaves already taken by the employee.'))
```

Comment:
This doesn't take into account the possibility of leave types allowing negative amounts
```suggestion
                if total_current_excess <= total_previous_excess:
                    continue
                lt = allocation.holiday_status_id 
                if lt.allows_negative and total_current_excess <= lt.max_allowed_negative:
                    continue 
                raise ValidationError(_('You cannot reduce the duration below the duration of leaves already taken by the employee.'))
```